File: clients/java/src/test/java/com/antzkv/TestMultithread.java
package com.antzkv;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

public class TestMultithread {
    static final String HOST = "127.0.0.1";
    static final int PORT = 6379;
    static final int THREADS = 10;
    static final int OPS = 100;

    public static void main(String[] args) throws Exception {
        testSingleClient();
        testPool();
    }

    /** One client shared across threads. */
    static void testSingleClient() throws Exception {
        AntzKVClient client = new AntzKVClient(HOST, PORT);
        client.connect();
        ExecutorService exec = Executors.newFixedThreadPool(THREADS);
        List<Future<?>> futures = new ArrayList<>();
        List<String> errors = new CopyOnWriteArrayList<>();

        for (int t = 0; t < THREADS; t++) {
            final int tid = t;
            futures.add(exec.submit(() -> {
                for (int i = 0; i < OPS; i++) {
                    String key = "t" + tid + "_k" + i;
                    String val = "v" + tid + "_" + i;
                    try {
                        client.set(key, val);
                        String got = client.get(key);
                        if (!val.equals(got)) {
                            errors.add("GET mismatch " + key + ": expected " + val + ", got " + got);
                        }
                        if (i % 10 == 0) client.delete(key);
                    } catch (Exception e) {
                        errors.add(key + " -> " + e.getMessage());
                    }
                }
            }));
        }
        for (Future<?> f : futures) f.get();
        exec.shutdown();
        client.close();
        System.out.println("[Single Client] Errors: " + errors.size());
        errors.stream().limit(5).forEach(e -> System.out.println("  " + e));
    }

    /** Connection pool test. */
    static void testPool() throws Exception {
        AntzKVClient.Pool pool = new AntzKVClient.Pool(HOST, PORT, 4);
        ExecutorService exec = Executors.newFixedThreadPool(THREADS);
        List<Future<?>> futures = new ArrayList<>();
        List<String> errors = new CopyOnWriteArrayList<>();
        long start = System.nanoTime();

        for (int t = 0; t < THREADS; t++) {
            final int tid = t;
            futures.add(exec.submit(() -> {
                for (int i = 0; i < OPS; i++) {
                    String key = "pool_t" + tid + "_k" + i;
                    String val = "pool_v" + tid + "_" + i;
                    try {
                        AntzKVClient c = pool.acquire();
                        try {
                            c.set(key, val);
                        } finally {
                            pool.release(c);
                        }
                    } catch (Exception e) {
                        errors.add(key + " -> " + e.getMessage());
                    }
                }
            }));
        }
        for (Future<?> f : futures) f.get();
        exec.shutdown();
        double elapsed = (System.nanoTime() - start) / 1e9;
        pool.close();
        int total = THREADS * OPS;
        System.out.printf("[Pool] %d ops in %.2fs = %.0f ops/sec | Errors: %d%n",
                          total, elapsed, total / elapsed, errors.size());
    }
}

File: clients/python/test_multithread.py
import threading
import time
from antzkv import AntzKVClient, AntzKVPool

HOST = "127.0.0.1"
PORT = 6379
NUM_THREADS = 10
OPS = 100

def test_single_client_multithread():
    """One client shared across threads."""
    client = AntzKVClient(HOST, PORT)
    client.connect()
    errors = []
    lock = threading.Lock()

    def worker(tid):
        for i in range(OPS):
            key = f"t{tid}_k{i}"
            val = f"v{tid}_{i}"
            if not client.set(key, val):
                with lock: errors.append(f"SET failed {key}")
            got = client.get(key)
            if got != val:
                with lock: errors.append(f"GET mismatch {key}: expected {val}, got {got}")
            if i % 10 == 0:
                client.delete(key)

    threads = [threading.Thread(target=worker, args=(t,)) for t in range(NUM_THREADS)]
    for t in threads: t.start()
    for t in threads: t.join()

    client.close()
    print(f"[Single Client] Errors: {len(errors)}")
    if errors:
        for e in errors[:5]: print("  ", e)

def test_pool():
    """Connection pool with concurrent threads."""
    pool = AntzKVPool(HOST, PORT, size=4)
    errors = []
    lock = threading.Lock()

    def worker(tid):
        for i in range(OPS):
            c = pool.acquire()
            try:
                key = f"pool_t{tid}_k{i}"
                val = f"pool_v{t}_{i}"
                if not c.set(key, val):
                    with lock: errors.append(f"SET failed {key}")
            finally:
                pool.release(c)

    threads = [threading.Thread(target=worker, args=(t,)) for t in range(NUM_THREADS)]
    start = time.time()
    for t in threads: t.start()
    for t in threads: t.join()
    elapsed = time.time() - start

    pool.close()
    print(f"[Pool] {NUM_THREADS * OPS} ops in {elapsed:.2f}s = {NUM_THREADS*OPS/elapsed:.0f} ops/sec | Errors: {len(errors)}")

if __name__ == "__main__":
    test_single_client_multithread()
    test_pool()
